### PR TITLE
feature(search): migrate discovery endpoints from entity-service to search-service

### PR DESCRIPTION
**Pull request by gpoujol**

### CODE REVIEW

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    id = toUri(it["attribute_name"]),
+                    attributeName = compactTerm(it["attribute_name"] as ExpandedTerm, contexts),
+                    attributeTypes = listOf(AttributeType.forKey(it["attribute_type"] as String))
+                )
+            }
+        ).right()
+    }
+
+    private fun rowToTypes(row: Map<String, Any>): List<ExpandedTerm> =
+        toList<ExpandedTerm>(row["types"]).filter { !AuthContextModel.IAM_TYPES.plus("Entity").contains(it) }
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt:95]: this is no longer needed:
- the `Entity` label was a neo4j thing to identify nodes holding entities
- users, clients and groups are no longer entities (look at `EntityAccessRight` and `SubjectReferential`)

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    private val databaseClient: DatabaseClient
+) {
+
+    suspend fun getEntityTypeList(contexts: List<String>): EntityTypeList {
+        val entityTypes = databaseClient.sql(
+            """
+            SELECT DISTINCT(types) FROM entity_payload
+            ORDER BY types
+            """.trimIndent()
+        ).allToMappedList { rowToTypes(it) }.flatten()
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt:31]: I'm wondering if you could have duplicates in your result.

For instance, if:
- Entity E1 has type Type T1 and Type T2
- Entity E2 has type Type T1

The distinct will return (T1, T2) and (T1), then you'll have T1, T2 and T1 in the final result?

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                    attributeTypes = listOf(AttributeType.forKey(it["attribute_type"] as String))
+                )
+            }
+        ).right()
+    }
+
+    private fun rowToTypes(row: Map<String, Any>): List<ExpandedTerm> =
+        toList<ExpandedTerm>(row["types"]).filter { !AuthContextModel.IAM_TYPES.plus("Entity").contains(it) }
+
+    private fun rowToEntityType(row: Map<String, Any>): List<Pair<String, String>> {
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt:97]: the result is `List<Pair<ExpandedTerm, ExpandedTerm>>`?

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                attributeNames = compactTerms(it.second, contexts).toSet().sorted()
+            )
+        }
+    }
+
+    suspend fun getEntityTypeInfoByType(
+        type: String,
+        contexts: List<String>
+    ): Either<APIException, EntityTypeInfo> {
+        val expandedType = expandJsonLdTerm(type, contexts)
```
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt:59]: it's expanded in the handler (and better to do the expansion in the handler)

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt (hunk 2 above)
Comment (gpoujol) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt:31]: The solution will to transform array to string or add toSet at the end

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt (hunk 3 above)
Comment (gpoujol) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt:97]: yes it's this

Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt:97]: so change the signature to make it clear

File: search-service/src/main/kotlin/com/egm/stellio/search/service/AttributeService.kt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        expandedAttribute: ExpandedTerm,
+        contexts: List<String>
+    ): Either<APIException, AttributeTypeInfo> {
+        val result = databaseClient.sql(
+            """
+            WITH entities AS (
+                SELECT entity_id, attribute_name, attribute_type
+                FROM temporal_entity_attribute 
+                WHERE :attribute = attribute_name
+                ORDER BY entity_id
```
[the remark this answers is not in the document]
Comment (gpoujol) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/AttributeService.kt:65]: it's not neccesary

File: search-service/src/test/kotlin/com/egm/stellio/search/service/EntityTypeServiceTests.kt [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    )
+
+    @AfterEach
+    fun clearPreviousTemporalEntityAttributesAndObservations() {
+        r2dbcEntityTemplate.delete(EntityPayload::class.java)
+            .all()
+            .block()
+        r2dbcEntityTemplate.delete(AttributeInstance::class.java)
+            .all()
+            .block()
```
Comment (bobeal) [inline search-service/src/test/kotlin/com/egm/stellio/search/service/EntityTypeServiceTests.kt:86]: I don't think you are creating things in `attribute_instance` table?

File: search-service/src/test/kotlin/com/egm/stellio/search/service/AttributeServiceTest.kt [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    )
+
+    @AfterEach
+    fun clearPreviousTemporalEntityAttributesAndObservations() {
+        r2dbcEntityTemplate.delete(EntityPayload::class.java)
+            .all()
+            .block()
+        r2dbcEntityTemplate.delete(AttributeInstance::class.java)
+            .all()
+            .block()
```
Comment (bobeal) [inline search-service/src/test/kotlin/com/egm/stellio/search/service/AttributeServiceTest.kt:85]: I don't think you are creating things in attribute_instance table?

Comment (gpoujol) [reply search-service/src/test/kotlin/com/egm/stellio/search/service/AttributeServiceTest.kt:85]: Yes I remove it

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt (hunk 2 above)
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt]: did you add a test for this use case?

Comment (gpoujol) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt]: When we create entity payload the first contains BEEHIVE_TYPE and SENSOR_TYPE, the second only SENSOR_TYPE, and the last APIARY_TYPE. When we do a get to retrieve type list I receive only 3 types. It's ok this test ?

Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityTypeService.kt]: LGTM

**Review by bobeal — APPROVED** [review]

**Pull request merged**